app/context.go

```diff
@@ -0,0 +1,79 @@
+package app
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+)
+
+type ContextKey string
+
+const (
+	TraceIDKey  ContextKey = "trace_id"
+	UserIDKey   ContextKey = "user_id"
+	TenantIDKey ContextKey = "tenant_id"
+)
+
+// Context carries the context of the current execution.
+type Context struct {
+	// original context
+	context.Context
+}
+
+// UserID returns the user id
+func (sc *Context) UserID() (string, bool) {
+	userID := sc.Context.Value(UserIDKey)
+	id, ok := userID.(string)
+	return id, ok
+}
+
+// SetUserID sets the user id
+func (sc *Context) SetUserID(userID string) *Context {
+	sc.Context = context.WithValue(sc.Context, UserIDKey, userID)
+
+	return sc
+}
+
+// TenantID returns the tenant id
+func (sc *Context) TenantID() (string, bool) {
+	tenantIDKey := sc.Context.Value(TenantIDKey)
+	id, ok := tenantIDKey.(string)
+	return id, ok
+}
+
+// SetTenantID sets the user id
+func (sc *Context) SetTenantID(tenantID string) *Context {
+	sc.Context = context.WithValue(sc.Context, TenantIDKey, tenantID)
+
+	return sc
+}
+
+// SetTraceID sets the trace id
+func (sc *Context) SetTraceID(traceID string) *Context {
+	sc.Context = context.WithValue(sc.Context, TraceIDKey, traceID)
+
+	return sc
+}
+
+// TraceID returns the trace identifier for the current flow
+func (sc *Context) TraceID() string {
+	traceID := sc.Context.Value(TraceIDKey)
+	id, ok := traceID.(string)
+
+	if !ok {
+		return ""
+	}
+
+	return id
+}
+
+// NewContext returns a new Context from a context.Context
+func NewContext(ctx context.Context) Context {
+	appCtx := Context{Context: ctx}
+
+	if _, ok := ctx.Value(TraceIDKey).(string); !ok {
```

Comment:
I would have called TraceID, and define one if the if was == "", my problem is that you somehow define two places where you check if you have a context.

This comment is made on your current code.

I think you could also consider to do not play/check/set with traceid in app.NewContext, but to define one when none is present in GetTraceID.

It would also fix an issue that someone with your current GetTraceID implementation.

Someone could define a context like this

```go

ctx : = app.Context{Context: context.Background()} // possible as app.Context cannot be nothing but Exported

traceID := ctx.TraceID() // this would be empty

```

Reply:
If you never call the `TraceID()` and pass the context as a value instead of a reference, the `TraceID` value will be mismatched.

Reply:
Indeed, but the points I raised are still valid, no?

Reply:
Yes, I added in the constructor and getter but it does not feel right 🤔